internal/logger/io_sink.go

```diff
@@ -7,91 +7,54 @@
+
+	// encMu protects the encoder from concurrent writes. While the logger
+	// itself does not concurrently write to the sink, the sink may be used
+	// concurrently within the driver.
+	encMu sync.Mutex
 }
 
-// Compile-time check to ensure osSink implements the LogSink interface.
+// Compile-time check to ensure IOSink implements the LogSink interface.
 var _ LogSink = &IOSink{}
 
-// NewIOSink will create a new IOSink that writes to the provided io.Writer.
+// NewIOSink will create an IOSink object that writes JSON messages to the
+// provided io.Writer.
 func NewIOSink(out io.Writer) *IOSink {
 	return &IOSink{
-		log: log.New(out, "", log.LstdFlags),
+		enc: json.NewEncoder(out),
 	}
 }
 
-func logCommandMessageStarted(log *log.Logger, kvMap map[string]interface{}) {
-	format := "Command %q started on database %q using a connection with " +
-		"server-generated ID %d to %s:%d. The requestID is %d and " +
-		"the operation ID is %d. Command: %s"
-
-	log.Printf(format,
-		kvMap["commandName"],
-		kvMap["databaseName"],
-		kvMap["serverConnectionId"],
-		kvMap["serverHost"],
-		kvMap["serverPort"],
-		kvMap["requestId"],
-		kvMap["operationId"],
-		kvMap["command"])
-
-}
-
-func logCommandMessageSucceeded(log *log.Logger, kvMap map[string]interface{}) {
-	format := "Command %q succeeded in %d ms using server-generated ID " +
-		"%d to %s:%d. The requestID is %d and the operation ID is " +
-		"%d. Command reply: %s"
+// Info will write a JSON-encoded message to the io.Writer.
+func (sink *IOSink) Info(_ int, msg string, keysAndValues ...interface{}) {
+	kvMap := make(map[string]interface{}, len(keysAndValues)/2+2)
 
-	log.Printf(format,
-		kvMap["commandName"],
-		kvMap["duration"],
-		kvMap["serverConnectionId"],
-		kvMap["serverHost"],
-		kvMap["serverPort"],
-		kvMap["requestId"],
-		kvMap["operationId"],
-		kvMap["reply"])
-}
-
-func logCommandMessageFailed(log *log.Logger, kvMap map[string]interface{}) {
-	format := "Command %q failed in %d ms using a connection with " +
-		"server-generated ID %d to %s:%d. The requestID is %d and " +
-		"the operation ID is %d. Error: %s"
+	kvMap[KeyTimestamp] = time.Now().UnixNano()
+	kvMap[KeyMessage] = msg
 
-	log.Printf(format,
-		kvMap["commandName"],
-		kvMap["duration"],
-		kvMap["serverConnectionID"],
-		kvMap["serverHost"],
-		kvMap["serverPort"],
-		kvMap["requestId"],
-		kvMap["operationId"],
-		kvMap["failure"])
-}
-
-func (osSink *IOSink) Info(_ int, msg string, keysAndValues ...interface{}) {
-	kvMap := make(map[string]interface{})
 	for i := 0; i < len(keysAndValues); i += 2 {
 		kvMap[keysAndValues[i].(string)] = keysAndValues[i+1]
 	}
 
-	switch msg {
-	case CommandStarted:
-		logCommandMessageStarted(osSink.log, kvMap)
-	case CommandSucceeded:
-		logCommandMessageSucceeded(osSink.log, kvMap)
-	case CommandFailed:
-		logCommandMessageFailed(osSink.log, kvMap)
+	sink.encMu.Lock()
+	defer sink.encMu.Unlock()
+
+	if err := sink.enc.Encode(kvMap); err != nil {
+		panic(err)
```

Comment:
The `IOSink` should not panic on a write/encode error. Logging is best-effort and we have no other way to notify the user about the error, so ignore it.